Treat amp-script AMP.setState as high trust

samouri opened the pull request:
**summary**
Addresses https://github.com/ampproject/amphtml/issues/24862

With this change, we allow all `AMP.setState` calls triggered within an amp-script to cause bind reevaluations as long as there was a recent (5s) user interaction.

**open question**
1. Is 5 seconds too long of a user interaction window? If so, how short should it be?

cc @choumx / @jridgewell

choumx commented on extensions/amp-script/0.1/amp-script.js:
```diff
@@ -640,7 +649,8 @@ export class SanitizerImpl {
           if (state) {
-            bind.setState(state, /* skipEval */ true, /* skipAmpState */ false);
+            const skipEval = !this.userActivationTracker_.isActive();
+            bind.setState(state, skipEval, /* skipAmpState */ false);
```
Minor: Can we add a simple comment explaining this? E.g. "Only evaluate (mutate the DOM) if there was a recent user gesture.".

samouri replied:
Done

choumx commented on extensions/amp-script/0.1/amp-script.js:
```diff
@@ -640,7 +649,8 @@ export class SanitizerImpl {
           if (state) {
-            bind.setState(state, /* skipEval */ true, /* skipAmpState */ false);
+            const skipEval = !this.userActivationTracker_.isActive();
+            bind.setState(state, skipEval, /* skipAmpState */ false);
```
If skipEval == true, let's `user().warn` that the `AMP.setState` only updated the page state since there was no backing user gesture.

samouri replied:
Done. Thanks for the improvement!

samouri commented:
I just modified a recently added test case which appears to fail in saucelabs.
Only some dom emulators (jsdom etc.) understand `Element.nonce`, whereas they all work correctly with `Element.getAttribute`, so I've modified the single line to see if it fixes saucelabs.

mattwomple commented:
> I just modified a recently added test case which appears to fail in saucelabs.
> Only some dom emulators (jsdom etc.) understand `Element.nonce`, whereas they all work correctly with `Element.getAttribute`, so I've modified the single line to see if it fixes saucelabs.

@samouri I'm appreciative of the `test-validator-integration.js` fix, but it was confusing to me that this was included in a very unrelated PR. Could you break out changes like this into their own PR in the future?

samouri commented:
>Could you break out changes like this into their own PR in the future?

I was actually planning to after verifying that it fixed the issue, but then this PR was merged really quickly! I should have made a note not to submit until I could split it out.

Thanks for the reminder of good engineering practice and I'll make sure to next time ⚡️.

choumx commented:
🙃

syedimran22 commented:
Hi @samouri  and @choumx  thank you for implementing AMP.setState in amp-script but whereas Amp.setState is setting the previous DOM data, not the updated ones which means it is giving one step previous data of the events.

samouri commented:
@syedimran22: im not sure what you mean. Can you open up a separate issue (bug report) with steps to reproduce the issue you are running into?

syedimran22 commented:
Sure @samouri I will open a new ticket with the steps.

mjwettergreen commented:
> Hi @samouri and @choumx thank you for implementing AMP.setState in amp-script but whereas Amp.setState is setting the previous DOM data, not the updated ones which means it is giving one step previous data of the events.

Can confirm, I'm also seeing an issue that presents like this. When I enable more advanced logging in console I can see that amp-script comes back with a message from the worker that the event is trusted, and then amp-bind evaluation runs immediately, and _then_ right after that I see the storage log for the AMP.setState() I fired. So it appears like bind evals on the state that exists right before you set state, but even if you use AMP.setState() later during your script 5s run window (tested with timeout) it never actually triggers a bind eval on it's own. It's the next user interaction that does that.

Is there a bug report for this yet? I can try and help generate a simple example, the one I'm working on is too complex to share.

samouri commented:
@mjwettergreen: no bug report yet AFAIK. 
If you are able to use your complex example to help create a minimal reproduction that would definitely be helpful!

mjwettergreen commented:
@samouri Actually, I was litterally about to file a bug report on it today (mouse over the submit button) and then an update distributed that fixed it. Seems to be working now. I did create a minimal example [here](https://nwdesignlabs.com/dev/amp-script-test.html) but the behavior no longer occurs. previously, when you clicked on the red button, the state updated in the background but the bound text didn't change.
